File: 0526/같이 눈사람 만들래?.py
import sys
input = sys.stdin.readline

N = int(input())
arr = sorted(list(map(int, input().split())))

left = 0
right = N - 1

tmp = abs(arr[left] + arr[right])

while left < right:
    a = arr[left] + arr[right]
    for i in range(left + 1, right - 1):
        for j in range(right - 1, i, -1):
            b = arr[i] + arr[j]
            if tmp > abs(a - b):
                tmp = abs(a - b)
            if a >= b:
                break
                
    if tmp == 0:
        break

    if left == right - 1:
        right -= 1
        left = 0
    else:
        left += 1

print(tmp)

File: 0526/같이 눈사람 만들래?.java
import java.util.*;
import java.io.*;

public class Main {

    static int N; // 눈덩이 수
    static int[] heights;
    static int answer = Integer.MAX_VALUE;
    
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        heights = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

        Arrays.sort(heights);

        outter: for (int e1=0; e1<N; e1++) {
            for (int e2=e1+1; e2<N; e2++) {

                int eHeight = heights[e1] + heights[e2];
                f(eHeight, e1, e2);
                if (answer == 0) {
                    break outter;
                }
            }
        }

        System.out.println(answer);
    }

    static void f(int eHeight, int e1, int e2) {
        int left = 0;
        int right = N - 1;
        int aHeight = 0;
        int minDiff = Integer.MAX_VALUE;
    
        while (left < right) {
            if (left == e1 || left == e2) {
                left++;
                continue;
            }
            if (right == e1 || right == e2) {
                right--;
                continue;
            }
    
            aHeight = heights[left] + heights[right];
            int diff = Math.abs(eHeight - aHeight);
            if (diff == 0) {
                answer = 0;
                break;
            }
            minDiff = Math.min(minDiff, diff);
    
            if (eHeight < aHeight) {
                right--;
            } else {
                left++;
            }
        }
        
        // System.out.println("minDiff = " + minDiff);
        answer = Math.min(answer, minDiff);
    }
}
